Propagate DDL Commands with 2PC

aamederen opened the pull request:
Fixes #513

This change modifies the DDL Propagation logic so that DDL queries
are propagated via 2-Phase Commit protocol. This way, failures during
the execution of distributed DDL commands will not leave the table in
an intermediate state.

The workflow of the successful case is this:
1. Open individual connections to all shard placements and send `BEGIN`
2. Send `SELECT worker_apply_shard_ddl_command(<shardId>, <DDL Command>)`
to all connections, one by one, in a serial manner.
3. Send `PREPARE TRANSCATION <transaction_id>` to all connections.
4. Sedn `COMMIT` to all connections.

Failure cases:
- If a worker problem occurs before sending of all DDL commands is finished, then
all changes are rolled back.
- If a worker problem occurs after all DDL commands are sent but not after
`PREPARE TRANSACTION` commands are finished, then all changes are rolled back.
However, if a worker node is failed, then the prepared transactions in that worker
should be rolled back manually.
- If a worker problem occurs during `COMMIT PREPARED` statements are being sent,
then the prepared transactions on the failed workers should be commited manually.
- If master fails before the first 'PREPARE TRANSACTION' is sent, then nothing is
changed on workers.
- If master fails during `PREPARE TRANSACTION` commands are being sent, then the
prepared transactions on workers should be rolled back manually.
- If master fails during `COMMIT PREPARED` or `ROLLBACK PREPARED` commands are being
sent, then the remaining prepared transactions on the workers should be handled manually.

marcocitus commented on src/backend/distributed/utils/multi_transaction.c:
```diff
@@ -347,3 +348,62 @@ ConnectionList(HTAB *connectionHash)
+ * accordingly.
+ */
+void
+OpenConnectionsToShardPlacements(uint64 shardId, HTAB *shardConnectionHash,
```
What do you think of moving this into a transaction/shard_transaction.c file?

In the future, we'll have a worker_transaction.c related to doing transactions across all workers (e.g. syncing metadata), which uses the 2PC functions from multi_transaction.c. It would be nice if we separated logic for doing transactions across shard placements into a separate file.

marcocitus replied:
The other functions like GetShardConnections are also candidates to go into that file.

marcocitus commented on src/backend/distributed/executor/multi_utility.c:
```diff
@@ -1045,88 +1056,203 @@ ExecuteDistributedDDLCommand(Oid relationId, const char *ddlCommandString)
+		 * (2PC).
+		 */
+
+		PrepareRemoteTransactions(shardPlacementConnectionList);
```
Should we honour multi_shard_commit_protocol here?

aamederen replied:
I had removed the multi_shard_commit_protocol reference because it seemed to me that all DDL propagations should have been done via 2PC. I put it back now.

marcocitus commented on src/backend/distributed/executor/multi_utility.c:
```diff
@@ -1045,88 +1056,203 @@ ExecuteDistributedDDLCommand(Oid relationId, const char *ddlCommandString)
+ * The resulting hash table maps shardIds to ShardConnection structs.
+ */
+static HTAB *
+OpenConnectionsToAllShardPlacements(List *shardIdList, char *relationOwner)
```
This logic seems fairly generic and could be added to shard_transaction.c.

marcocitus commented on src/backend/distributed/executor/multi_utility.c:
```diff
@@ -1045,88 +1056,203 @@ ExecuteDistributedDDLCommand(Oid relationId, const char *ddlCommandString)
-	List *shardList = NIL;
+	List *shardIdList = LoadShardList(relationId);
 	char *relationOwner = TableOwner(relationId);
+	HTAB *shardConnectionHash = OpenConnectionsToAllShardPlacements(shardIdList,
```
If we were to implement multi-statement DDL changes. Would we want to store this hash in a static variable, or rebuild the hash from shardPlacementConnectionList?

aamederen replied:
One of doing that came to my mind is the following: We can make `shardConnectionsHash` an extern variable and make `OpenTransactionsToAllShardPlacements` aware of the current state of `shardConnectionsHash`. This way, we can have multiple calls to `ExecuteCommandOnWorkerShards`. What do you think?

marcocitus commented on src/backend/distributed/executor/multi_utility.c:
```diff
@@ -1045,88 +1056,203 @@ ExecuteDistributedDDLCommand(Oid relationId, const char *ddlCommandString)
+		PGresult *result = NULL;
+
+		/* send the query */
+		result = PQexec(connection, "BEGIN");
```
I think it might make more sense to send the BEGIN after opening the connection. It creates a bit more symmetry between open+begin and commit+close which will help when we add support for multi-statement transactions.

aamederen replied:
Right, it now sends the BEGIN after openin the connections in shard_transaction.c::OpenTransactionsToAllShardPlacements function.

After this change, BEGIN commands are now sent before PG_TRY in master_modify_multiple_shards.c::SendQueryToShards(), would it bring any problems?

marcocitus commented on src/backend/distributed/master/master_modify_multiple_shards.c:
```diff
@@ -266,81 +265,22 @@ static HTAB *
 OpenConnectionsToAllShardPlacements(List *shardIntervalList)
```
How similar is this function to the one in multi_utility.c? Would it be possible to unify them?

aamederen replied:
You are right, they were almost identical. After the change of sending BEGIN after opening the connections, it is also renamed to `OpenTransactionsToAllShardPlacements`

marcocitus commented on src/backend/distributed/executor/multi_utility.c:
```diff
@@ -1045,88 +1056,203 @@ ExecuteDistributedDDLCommand(Oid relationId, const char *ddlCommandString)
+
+	MemoryContextSwitchTo(oldContext);
+
+	if (!isXactCallbackRegistered)
```
For integration purposes, maybe move this if-block into an EnableXactCallback function (or better name).

e.g.
https://github.com/citusdata/citus-mx/blob/4c70aba5b300be28bb3a81187591246ef16b390f/src/backend/distributed/transaction/worker_transaction.c#L200

aamederen replied:
Moved the logic to a RegisterShardPlacementXactCallback function.

marcocitus commented on src/backend/distributed/utils/multi_transaction.c:
```diff
@@ -347,3 +348,62 @@ ConnectionList(HTAB *connectionHash)
+															 shardId,
+															 &shardConnectionsFound);
+
+	List *shardPlacementList = FinalizedShardPlacementList(shardId);
```
It seems like we never lock the metadata for the shards. Not this PR's fault, but this could cause things to break in the event that shard rebalancer runs during a DDL command.

aamederen replied:
Added `lockShards` to `ExecuteCommandOnWorkerShards`, before opening the connections and used ShareLock. Is it sufficient?

marcocitus commented:
I think it's worth adding regression tests for #480 at least.

marcocitus commented on src/include/distributed/shard_transaction.h:
```diff
@@ -0,0 +1,20 @@
+ *
+ *-------------------------------------------------------------------------
+ */
+
```
#ifndef is missing

aamederen replied:
Added.

marcocitus commented on src/backend/distributed/transaction/shard_transaction.c:
```diff
@@ -0,0 +1,212 @@
+		if (PQresultStatus(result) != PGRES_COMMAND_OK)
+		{
+			WarnRemoteError(connection, result);
+			ereport(ERROR, (errmsg("could not send ddl command to shard placement")));
```
Maybe just ReraiseRemoteError. If this function has another caller, the error message doesn't make much sense.

aamederen replied:
Makes sense

marcocitus commented on src/backend/distributed/transaction/shard_transaction.c:
```diff
@@ -0,0 +1,212 @@
+ * The resulting hash table maps shardIds to ShardConnection structs.
+ */
+HTAB *
+OpenTransactionsToAllShardPlacements(List *shardIdList, char *relationOwner)
```
relationOwner seems very specific to DDL, maybe just something like userName or nodeUser.

aamederen replied:
Changed nodeUser and relationOwner to userName.

marcocitus commented on src/backend/distributed/executor/multi_utility.c:
```diff
@@ -1045,88 +1055,175 @@ ExecuteDistributedDDLCommand(Oid relationId, const char *ddlCommandString)
+ * transactions when the local transaction commits or aborts.
+ */
+static void
+CompleteShardPlacementTransactions(XactEvent event, void *arg)
```
I wonder whether it makes sense to move this logic into shard_transaction.c as well, since other callers might want to use it.

marcocitus commented on src/backend/distributed/executor/multi_utility.c:
```diff
@@ -1045,88 +1055,175 @@ ExecuteDistributedDDLCommand(Oid relationId, const char *ddlCommandString)
+ * be removed by either COMMIT PREPARED or ROLLBACK PREPARED.
  */
 static bool
 ExecuteCommandOnWorkerShards(Oid relationId, const char *commandString,
```
failedPlacementList seems no longer used.

aamederen replied:
Correct, removed the failedPlacementList related logic as well.

marcocitus commented on src/backend/distributed/executor/multi_utility.c:
```diff
@@ -1045,88 +1055,175 @@ ExecuteDistributedDDLCommand(Oid relationId, const char *ddlCommandString)
-			ShardPlacement *placement = (ShardPlacement *) lfirst(shardPlacementCell);
-			char *workerName = placement->nodeName;
-			uint32 workerPort = placement->nodePort;
+		ExecuteCommandOnShardReplicas(applyCommand, shardId, shardConnections);
```
I guess we usually say Placements instead of Replicas in code.

aamederen replied:
You are right, changed its name.

marcocitus commented on src/backend/distributed/executor/multi_utility.c:
```diff
@@ -1045,88 +1055,175 @@ ExecuteDistributedDDLCommand(Oid relationId, const char *ddlCommandString)
+		isXactCallbackRegistered = true;
+	}
+	
+	foreach(shardCell, shardIdList)
```
If we wanted to parallelise this block, how would it change?

aamederen replied:
I am not sure but, I think using PQsendQuery instead of PQexec in ExecuteCommandOnShardPlacements function,  and checking the results after the foreach with `PQgetResult` would be needed.

aamederen replied:
Would you like me to try it in this PR?

aamederen commented:
> * What do you think of moving this into a transaction/shard_transaction.c file?
 In the future, we'll have a worker_transaction.c related to doing transactions across all workers (e.g. syncing metadata), which uses the 2PC functions from multi_transaction.c. It would be nice if we separated logic for doing transactions across shard placements into a separate file.

> * The other functions like GetShardConnections are also candidates to go into that file.

> * This logic seems fairly generic and could be added to shard_transaction.c.

> * I wonder whether it makes sense to move this logic into shard_transaction.c as well, since other callers might want to use it.

I have created `src/backend/distributed/transaction` directory, created `shard_transaction.c` file and also moved `multi_transaction.c` there. I have moved these functions to the `shard_transaction.c` file. 

However, I am not sure about what should be the distinction between shard_transaction.c and multi_transaction.c. 

Currently, they are separated so that `multi_transaction.c` has prepare/commit/abort related functions and `shard_transaction.c` has shard/placement connection-related functions. 

1. Do you think is this a good separation?
2. To me, the names of `shard_transaction.c` and `multi_transaction.c` do not tell any difference between them.
3. I think `CompleteShardPlacementTransactions` function may go to multi_transaction instead of shard_transaction, since it couples with the functions in multi_transaction.

marcocitus commented:
> To me, the names of shard_transaction.c and multi_transaction.c do not tell any difference between them.

Given that multi_transaction.c deals with generic 1PC/2PC operations. An alternative name could be commit_protocol.c

marcocitus commented on src/backend/distributed/executor/multi_utility.c:
```diff
@@ -1045,88 +1036,136 @@ ExecuteDistributedDDLCommand(Oid relationId, const char *ddlCommandString)
+	shardPlacementConnectionList = ConnectionList(shardConnectionHash);
+	MemoryContextSwitchTo(oldContext);
+
+	RegisterShardPlacementXactCallback();
```
Maybe move this into OpenTransactionsToAllShardPlacements.

aamederen replied:
Doing so will also change the master_modify_multiple_shards, since it uses PG_TRY after OpenTransactionsToAllPlacements. 

Would you prefer changing master_modify_multiple_shards error handling logic in this PR anyways or leave it to a separate technical dept issue.

aamederen replied:
Moved it inside the OpenTransactionsToAllShardPlacements and modified master_modify_multiple_shards accordingly.

marcocitus commented on src/backend/distributed/executor/multi_utility.c:
```diff
@@ -1045,88 +1036,136 @@ ExecuteDistributedDDLCommand(Oid relationId, const char *ddlCommandString)
+	shardConnectionHash = OpenTransactionsToAllShardPlacements(shardIntervalList,
+															   tableOwner);
+
+	shardPlacementConnectionList = ConnectionList(shardConnectionHash);
```
Maybe move this into OpenTransactionsToAllShardPlacements

aamederen replied:
Moved

marcocitus commented on src/backend/distributed/executor/multi_utility.c:
```diff
@@ -1045,88 +1036,136 @@ ExecuteDistributedDDLCommand(Oid relationId, const char *ddlCommandString)
-										" on node %s:%d", shardId, workerName,
-										workerPort)));
-			}
+	/* CloseConnections(allShardsConnectionList); */
```
Should this be removed?

aamederen replied:
Yeah, sorry for my overlook, removed now.

marcocitus commented on src/backend/distributed/transaction/shard_transaction.c:
```diff
@@ -0,0 +1,279 @@
+ * The resulting hash table maps shardIds to ShardConnection structs.
+ */
+HTAB *
+OpenTransactionsToAllShardPlacements(List *shardIntervalList, char *userName)
```
We usually try to follow calling order in our files, so this function should be above CreateShardConnectionHash.

aamederen replied:
You are right, fixed the file.

marcocitus commented on src/backend/distributed/transaction/shard_transaction.c:
```diff
@@ -0,0 +1,279 @@
+HTAB *
+OpenTransactionsToAllShardPlacements(List *shardIntervalList, char *userName)
+{
+	HTAB *shardConnectionHash = CreateShardConnectionHash();
```
The current code allows transaction blocks and gets into a deadlock because it opens new connections for every statement.

    postgres=# BEGIN;
    BEGIN
    postgres=# ALTER TABLE test ADD COLUMN z int;
    ALTER TABLE
    postgres=# ALTER TABLE test ADD COLUMN q int;
    ...

What if we just go all the way by caching the shardConnectionHash instead of shardPlacementConnectionList and returning it here if it is not null? That would allow multi-statement DDL transactions.

aamederen replied:
Put a `PreventTransactionChain` call in `ExecuteDistributedDDLCommand` function to error out in this case.

marcocitus commented:
You might want to check if there's any relevant overlap with #627, e.g. in the Xact handler.

marcocitus commented on src/backend/distributed/master/master_modify_multiple_shards.c:
```diff
@@ -198,10 +198,12 @@ HasReplication(List *shardIntervalList)
 	int affectedTupleCount = 0;
-	HTAB *shardConnectionHash = OpenConnectionsToAllShardPlacements(shardIntervalList);
+	char *relationOwner = TableOwner(relationId);
+	HTAB *shardConnectionHash = OpenTransactionsToAllShardPlacements(shardIntervalList,
```
I just realised that if you move RegisterShardPlacementXactCallback() into OpenTransactionsToAllShardPlacements, the behaviour would change here, but what if we remove the PG_TRY/PG_CATCH block now?

aamederen commented:
> Given that multi_transaction.c deals with generic 1PC/2PC operations, which could be across all shards or across all workers, an alternative name could be commit_protocol.c
> 
> Perhaps shard_transaction.c should be multi_shard_transaction.c for clarity.

Makes sense, changed accordingly and pushed a commit.

marcocitus commented on src/backend/distributed/transaction/multi_shard_transaction.c:
```diff
@@ -0,0 +1,303 @@
+/*-------------------------------------------------------------------------
+ *
+ * shard_transaction.c
```
should be multi_shard_transaction.c

aamederen replied:
fixed

marcocitus commented on src/include/distributed/multi_shard_transaction.h:
```diff
@@ -0,0 +1,44 @@
+/*-------------------------------------------------------------------------
+ *
+ * shard_transaction.h
```
should be multi_shard_transaction

aamederen replied:
Fixed

marcocitus commented:
Just ran some tests. Looks good overall apart from the multi-statement issue.

The 1PC failure scenario is not so nice right now:

    postgres=# ALTER TABLE test ADD COLUMN foo int;
    WARNING:  failed to commit transaction on localhost:9700
    WARNING:  failed to commit transaction on localhost:9700
    ALTER TABLE
    postgres=# SELECT * FROM test;
    WARNING:  column "foo" does not exist
    CONTEXT:  while executing command on localhost:9700
    WARNING:  column "foo" does not exist
    CONTEXT:  while executing command on localhost:9700
    WARNING:  column "foo" does not exist
    CONTEXT:  while executing command on localhost:9700
    WARNING:  column "foo" does not exist
    CONTEXT:  while executing command on localhost:9700
    WARNING:  column "foo" does not exist
    CONTEXT:  while executing command on localhost:9700
    ERROR:  failed to execute job 8
    DETAIL:  Failure due to failed task 6

We could provide better instructions on what to do after a failure, though maybe it's better if we enforce 2PC for DDL changes (e.g. error out if commit_protocol is not 2pc).

aamederen commented:
> We could provide better instructions on what to do after a failure, though maybe it's better if we enforce 2PC for DDL changes (e.g. error out if commit_protocol is not 2pc).

At the beginning, I though about, regardless of multi_shard_commit_protocol, we should do it in 2PC. 

I think now that instead of doing 2PC silently, forcing user to `SET citus.multi_shard_commit_protocol TO '2pc'` makes more sense. However, this would require changes in more than 10 regression tests.

aamederen commented:
Went back to using 2PC silently, even the `citus.multi_shard_commit_protocol` is set to 1pc.

marcocitus commented on src/backend/distributed/executor/multi_utility.c:
```diff
@@ -1044,89 +1039,148 @@ ExecuteDistributedDDLCommand(Oid relationId, const char *ddlCommandString)
+ * is '1pc';
+ */
+static void
+SwitchTo2PCForSession()
```
It's only the transaction because of GUC_ACTION_LOCAL.

aamederen replied:
Yeah, changed accordingly.

marcocitus commented on src/backend/distributed/master/master_modify_multiple_shards.c:
```diff
@@ -162,7 +163,7 @@ LockShardsForModify(List *shardIntervalList)
 		lockMode = ShareLock;
 	}
-	else if (!HasReplication(shardIntervalList)) /* check if any shards have >1 replica */
+	else if (!HasReplication(shardIntervalList)) /* check if any shard has >1 replicas */
```
We don't usually put code comments at the end of the line, should be on the line above.

aamederen replied:
Removed the comment, since the line is self-explanatory

marcocitus commented on src/backend/distributed/transaction/commit_protocol.c:
```diff
@@ -197,6 +260,8 @@ CommitRemoteTransactions(List *connectionList, bool stopOnFailure)
 				}
 			}
+
+			ereport(DEBUG2, (errmsg("sent COMMIT PREPARED over connection %ld", connectionId)));
```
are the connection ids meaningful to a user / other developer?

marcocitus replied:
If we want to show something here, it might be nice to just show the command that was sent, such that the user can see the transactions that were prepared/committed.

aamederen replied:
no, but I though that distinguishing between matching commits/rollbacks wouldn't hurt anyone.

marcocitus commented on src/test/regress/output/multi_alter_table_statements.source:
```diff
@@ -412,6 +412,126 @@ ERROR:  renaming distributed tables or their objects is currently unsupported
+DETAIL:  Key (a)=(2) is duplicated.
+CONTEXT:  while executing command on localhost:57638
+ERROR:  could not execute DDL command on worker node shards
+SELECT * FROM pg_dist_shard_placement WHERE shardstate=3;
```
This query seems like it would break if any other tests produce shardstate 3 in the future.

marcocitus commented on src/test/regress/output/multi_alter_table_statements.source:
```diff
@@ -412,6 +412,126 @@ ERROR:  renaming distributed tables or their objects is currently unsupported
+(1 row)
+
+DROP INDEX temp_index_1;
+-- verify that distributed ddl commands are not allowd inside a transaction block
```
s/allowd/allowed/

marcocitus commented on src/backend/distributed/executor/multi_utility.c:
```diff
@@ -77,9 +83,12 @@ static void ErrorIfDistributedRenameStmt(RenameStmt *renameStatement);
-										 List **failedPlacementList);
+static void ExecuteDistributedDDLCommand(Oid relationId, const char *ddlCommandString,
+										 bool isTopLevel);
+static void SwitchTo2PCForTransaction();
```
Should have void between the brackets.

This function is the equivalent of SET LOCAL citus.multi_shard_commit_protocol TO '2pc'. I think it might make sense to make that more explicit in the name, e.g. SetLocalCommitProtocolTo2PC.

marcocitus commented:
Apart from naming issue and re-running make reindent, this seems ready to 🚢 .

lithp commented on src/backend/distributed/transaction/commit_protocol.c:
```diff
@@ -47,6 +46,66 @@ InitializeDistributedTransaction(void)
+	{
+		/*
+		 * A failure here will cause some remote changes to either
+		 * roll back (1PC) or, in case of connection or node failure,
```
Is this comment wrong? Since you're giving `CommitRemoteTransactions` the false flag nothing will be rolled back upon error?
